astropy/io/fits/connect.py

```diff
@@ -181,6 +181,11 @@ def read_table_fits(input, hdu=None, astropy_native=False, memmap=False,
                 raise ValueError(f"No table found in hdu={hdu}")
 
         elif len(tables) == 1:
+            if not (hdu is None or hdu in tables):
+                warnings.warn(f"No table found in specified hdu={hdu},"
+                              " reading in first available "
+                              f"table (hdu={first(tables)}) instead!",
+                              AstropyUserWarning)
```

Comment:
You need a more strict test because currently `hdu in tables` gives unexpected results (I think there is a ticket about this). `hdu=5` or `hdu=''` will not raise the warning.

Reply:
Is that an `OrderedDict` issue? `hdu=''` wouldn't have surprised me too much, but `5`?
I cannot reproduce this with any Python 3.6-3.8 on MacOS X, and adding `5` and `''` to the tests still works, but if it fails on some platforms... what's a safe method for testing then? Would not need an `OrderedDict` in fact, since the standard dictionary already preserves order on 3.6+.

Reply:
It is ordered by insertion. Is it possible that insertion order is random?

Reply:
It's ordered by `enumerate(input)`, but `input` here is a `HDUList`, so if that has disorder issues that could be a problem.

Reply:
Oh sorry I missed that `tables` here is an OrderedDict, not the HDUList... forgot what I said :/

Reply:
Hmm, maybe not a problem with the order then? But I don't know what the comment about  lazy loading means, maybe @saimn knows.

https://github.com/astropy/astropy/blob/e2d877ec06745979b844a8d82bc137204726ae72/astropy/io/fits/hdu/hdulist.py#L239-L249

Reply:
This is the case that should not raise a warning (with a string argument):
```
In [12]: Table.read('astropy/io/fits/tests/data/memtest.fits', hdu='AXAF_CCDM')                
WARNING: No table found in specified hdu=AXAF_CCDM, reading in first available table (hdu=1) instead! [astropy.io.fits.connect]
Out[12]: 
<Table length=1>
       TIME       TLM_FMT  MJF   MNF   QUALITY [69]  ... CXPNAMOD CXPNARNG CXPNBMOD CXPNBRNG
        s                                            ...                                    
     float64       uint8  int32 int32      bool      ...  bytes4   bytes3   bytes4   bytes3 
----------------- ------- ----- ----- -------------- ... -------- -------- -------- --------
80348638.04702236       2  6887     0 False .. False ...     COHO       ON     COHO       ON
```

Reply:
@pllim @dhomeier - Sorry for the confusion, there is no ordering issue (unless I misunderstood something), it was just because of me believing that `tables` was the `HDUList`.
@pllim - In the `__iter__` loop the goal is to avoid loading all the HDUs when only one is requested, and for this `len` must not be called.

Reply:
Ok, ok. Ignore me then. Sorry for the noise.

Reply:
> In [12]: Table.read('astropy/io/fits/tests/data/memtest.fits', hdu='AXAF_CCDM')                
> WARNING: No table found in specified hdu=AXAF_CCDM, reading in first available table (hdu=1) instead! [astropy.io.fits.connect]

Strange, that's exactly the case that should be covered e.g. by `test_read_with_hdu_1['first']`. Only with multiple tables, but I just checked that this passes, too:
```
    @pytest.mark.parametrize('hdu', [None, 1, 'first'])
    def test_read_from_single_hdu(self, hdu):
        with catch_warnings() as l:
            t = Table.read(self.hdus[1])
        assert len(l) == 0
        assert equal_data(t, self.data1)
```

Reply:
I stand corrected; the test above worked reading directly from one `HDU`, not a `HDUList` with a single valid item. I can reproduce the failure now; will have to specifically check each `hdu.name` indeed.